### ISSUE TITLE
dynamic labels

### ISSUE THREAD

**Issue by msaarna** [dasm-assembler/dasm#42]
Dasm doesn't have the ability to create programmatic/dynamic labels. (edit: outside of macro arguments being used in labels, but those can't evaluate another symbol) I've created a [new feature branch](https://github.com/dasm-assembler/dasm/tree/feature/dynamic_labels) to address this shortcoming. See the [snapshot directory](https://github.com/dasm-assembler/dasm/tree/feature/dynamic_labels/snapshot-builds) if you want to play along at home.

I've added a "," operator to the label syntax, where a comma indicates that a symbol or string literal will follow the label (stub).

e.g.
IDX SET 10
MyDynamicLabel1_,IDX,"_X"    ; this will create a new label symbol "MyDynamicLabel1_10_X"
MyDynamicLabel2_,IDY,"_Y"    ; in the first pass, no label will be defined, since IDY is undefined
IDY SET 20 ; in this case, the second pass will define the label "MyDynamicLabel2_20_Y"
 ; undefined dynamic labels will force an additional pass, and eventually cause compilation to fail if they don't resolve

As it stands, this is very useful for programmatic data tables. There seems to be a bit of a use-case gap for using your programmatic labels as table data.  This can be worked around by placing the data generating command in a macro and using macro argument substitution to create the data... I can't think of a less kludgy implementation for that, so it may stay that way.

**Comment by msaarna** [comment]
I had to back off on a label component being undefined causing another pass. This was breaking assembly when dynamic labels were inside unused macros and inactive-conditionals. 
The partially defined dynamic label still doesn't get added to the symbol table, so anything referencing it will cause additional passes anyway.
Fixed in the dynamic_labels repo, and the snapshot builds therein.

**Comment by andrew-davie** [comment]
I was hoping to use this, but what i really need is the ability to print the string text name of a label, rather than the value of it. I want to use it to echo where an error occurs in a "anonymous" function

In particular I bracket banks with

 ROMBANK NAME
 CHECK_BANK_SIZE "NAME"

The check bank size will print the name if there's an issue, so I can easily see where it is.
I want to avoid having to type in "NAME" there, and instead have teh ROMBANK macro somehow store NAME --> "NAME" or in some way to be able to convert it.

I was thinking of using dynamic labels to create a new label, but I ran into the issue of requiring a quote at the very start, so couldn't see how I could use a dynamic label code to generate a quoted string, using the label name

Any ideas?

**Comment by msaarna** [comment]
You can convert labels to label strings with the recently added SETSTR pseudo-op...

   MAC SIZEOF
.NAME SETSTR {1}
        echo " The Size of ",.NAME,"is:",[* - {1} ]d,"bytes."
   ENDM

**Comment by andrew-davie** [comment]
Many thanks. That worked nicely in one situation I needed.
I still have another issue.


In this minimal example, and forgive me I’m tired…


    MAC DEF
{1}
FN SETSTR {1}
    ENDM

    MAC VEND
VEND_,FN SET *
    ENDM



    DEF Here
    VEND



I’m trying to “DEFINE” something using the Mac DEF
It’s passed the name of a routine to define.

So, first line it simply plonks down the label via {1}
The next line copies the text name via SETSTR to the label FN

Next the macro VEND
This is trying to merge the text “VEND_” and the contents of FN (which will be “Here”) to produce a label

VEND_Here SET *

This doesn’t seem to work well at all.
Have I misunderstood usage of dynamic labels here?

Thanks
A

**Comment by msaarna** [comment]
Nope, you have it right, it's just my implementation is apparently half-baked. Right now only stand-alone labels are being processed correctly, so the pseudo-op or op after the dynamic label is messed up. I also see another issue where the string set in one macro is being taken as a value in another macro.

I'll need to work through this. Thanks for the new test cases.

**Comment by andrew-davie** [comment]
I can't seem to get dynamic labels to work in the case where the value I want to "concatenate" is a string. In particular, I want to do something like this...

```NAME SETSTR TEST
,DYNAMIC SET 1```

In the above case, I'd like the label "TESTDYNAMIC" to have the value of 1

I simply get a segment fault when I try the above, and if instead I try...

```X,DYNAMIC SET 1```

Then I get an actual symbol with the name "X,DYNAMIC" instead of the desired "XTEST"

Am I misunderstanding how this works?

**Comment by andrew-davie** [comment]
I can't seem to get dynamic labels to work in the case where the value I want to "concatenate" is a string. In particular, I want to do something like this...

```
NAME SETSTR TEST
,DYNAMIC SET 1
```
In the above case, I'd like the label "TESTDYNAMIC" to have the value of 1
I simply get a segment fault when I try the above, and if instead I try...

```
X,DYNAMIC SET 1
```
Then I get an actual symbol with the name "X,DYNAMIC" instead of the expected "XTEST"

Am I misunderstanding how this works?

**Comment by andrew-davie** [comment]
I can't seem to get dynamic labels to work in the case where the value I want to "concatenate" a string. In particular, I want to do something like this...

```
NAME SETSTR TEST
NAME,DYNAMIC SET 1
```
In the above case, I'd like the label "TESTDYNAMIC" to have the value of 1
However, I just get a label in the symbol table "NAME,DYNAMIC" --yes, with a comma!

What I actually want to do, though, is....

```
NAME SETSTR TEST
,NAME SET 1
```
... and have the label "TEST" set to 1
Am I misunderstanding how this works?

**Comment by msaarna** [comment]
The first part of the dynamic label is a traditional label base, with the stuff following any comma arguments being evaluated as the dynamic bit. Also, In your example, if you want the DYNAMIC text to be literally in the label, you need to quote it. As it is, it will be evaluated as a symbol.

Are you sure you're using the dynamic-labels branch for your test? If the components of a dynamic label don't fully resolve, the symbol shouldn't be added to the table. If I run your example, this is what happens for me; I don't get a literal "NAME,DYNAMIC" added to the symbol table. 

All of that said, I do see a problem where dynamic labels isn't handling string constants correctly. It's using the numbered value of the string, rather than expressing it as a string. I expect I need to add a special case for string evaluations.

So in summary, this works as expected...
NAME SET 42
BASE,NAME,"DYNAMIC" SET 1
...creating the symbol table entry: "BASE42DYNAMIC            0001"

But this doesn't (yet) work as expected due to dynamic labels not handling string symbols correctly...
NAME SETSTR TEST
BASE,NAME,"DYNAMIC" SET 1
...incorrectly creating the symbol table entry: "BASE1413829460DYNAMIC    0001"

**Comment by andrew-davie** [comment]
Thanks for the quick reply.

> The first part of the dynamic label is a traditional label base, with the stuff following any comma arguments being evaluated as the dynamic bit.

Yes, I figured that - this is why I gave the two examples.
In actual fact I really need the "first part" to be optional. I already have the full name of a label I want to use - but it's a string. I was hoping to use "dynamic labels" to let me use the string as a label.

> Also, In your example, if you want the DYNAMIC text to be literally in the label, you need to quote it. As it is, it will be evaluated as a symbol.

Noted, but this was an error in my example, not the actual problem I am trying to solve.

> Are you sure you're using the dynamic-labels branch for your test?

No. Probably a "me bad" there. I am for sure now, though.

> So in summary, this works as expected...
> NAME SET 42
> BASE,NAME,"DYNAMIC" SET 1
> ...creating the symbol table entry: "BASE42DYNAMIC 0001"

That is the use-case I do not use/need.

> But this doesn't (yet) work as expected due to dynamic labels not handling string symbols correctly...
> NAME SETSTR TEST
> BASE,NAME,"DYNAMIC" SET 1
> ...incorrectly creating the symbol table entry: "BASE1413829460DYNAMIC 0001"

That is almost the use-case I need. I need "BASE" in this case to be a null string, or allow the syntax without it present (that is ",NAME SET 1")

What I'm trying to do is further-automate having dasm calculate local variables. That is, variables that are declared with a function (instead of in a zero page file), and use a common "overlay" area. I have this all working, but I need to put a superfluous "end of variables" macro in and duplicate the name of the subroutine as a parameter. If I can save the name of the subroutine in the first macro (which defines the subroutine), then I can re-use that name at the end of the variable block without having to re-type it.  It's complex but something that's very desirable.

**Comment by andrew-davie** [comment]
Could we please migrate this branch to master, as it's really useful and now tested in anger.

**Comment by msaarna** [comment]
I'll have a look at the latest request, with the first part of the label being dynamic too (or at least null), and we can see if it satisfies your use-cases. This branch was made to keep experiments out of master, which I think is a good thing, but once we're past experimental status I'm good with merging it in.

**Comment by andrew-davie** [comment]
Since we have another pull request in the works - even though it needs some work - I am thinking it is a good time to review this one and get it into master sooner rather than later. I've been using it for some time now, with no issues on complex codebases.

**Comment by dionoid** [comment]
@msaarna Hi Mike, are you okay if we merge the dynamic-labels branch into master, as it seems both you and Andrew did a lot of testing already?
Ideally, the merge would also include changes to /doc/dasm.txt with a short explanation of the feature.

**Comment by msaarna** [comment]
@andrew-davie and @dionoid  - Sorry. I had intended to reply back here, and work the code a bit further per Andrew's previous feedback, but I'm just completely under water in real life.
I'm good with merging the code back as-is. I think there's some deficiencies that Andrew identified above, but the code is useful enough as-is, and I'm not worried it will break anything, nor am I worried that addressing those deficiencies later will break anything.
The doc/dasm.txt in this branch already has a description of the feature.

**Comment by andrew-davie** [comment]
Extremely useful addition to the capabilities, and I use it extensively.
It would be nice to have some record/track of the non-implemented stuff so that it doesn't get forgotten?

**Comment by Settis** [comment]
It seems that the dynamic labels feature was released already. Is it makes sense to close this issue?

**Comment by patricksurry** [comment]
When using dynamic labels inside macros, it seems like a macro argument can only be used as the first part of a dynamic label.  Is that intended/documented? For example this works:
```
index set #1
  MAC FOO
    ;  FOO label
{1},"_test_",index
```
but switching to `test_,test_index,{1}` fails with an uninformative error:
```
word16.asm (499): error: Illegal character '{1}'.
word16.asm (499): error: Illegal character '}'.
```

**Comment by msaarna** [comment]
I pushed an update to master regarding dynamic labels. You should be able to use {#} as an argument in a dynamic label now. The doc could maybe be a bit clearer, but the start of the label isn't the dynamic part, the arguments are.  When used as a dynamic label argument, strings won't be quoted. 

For testing, you can try the automated snapshot build [here](https://github.com/dasm-assembler/dasm/actions/runs/5780124401)

**Comment by msaarna** [comment]
I'm going to close this one, since the original issue was resolved long ago. If there's any remaining issue, please open a new one.
